x-pack/plugin/core/src/main/java/org/elasticsearch/license/LicenseService.java

```diff
@@ -353,17 +354,32 @@ protected ClusterStateUpdateResponse newResponse(boolean acknowledged) {
                 @Override
                 public ClusterState execute(ClusterState currentState) throws Exception {
                     MetaData metaData = currentState.metaData();
-                    final LicensesMetaData currentLicenses = metaData.custom(LicensesMetaData.TYPE);
-                    if (currentLicenses.getLicense() != LicensesMetaData.LICENSE_TOMBSTONE) {
+                    final LicensesMetaData currentLicensesMetadata = metaData.custom(LicensesMetaData.TYPE);
+                    final License currentLicense = LicensesMetaData.extractLicense(currentLicensesMetadata);
+                    if (shouldRemove(currentLicense)) {
                         MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
-                        LicensesMetaData newMetadata = new LicensesMetaData(LicensesMetaData.LICENSE_TOMBSTONE,
-                            currentLicenses.getMostRecentTrialVersion());
-                        mdBuilder.putCustom(LicensesMetaData.TYPE, newMetadata);
+                        License.Builder specBuilder = License.builder()
+                            .uid(UUID.randomUUID().toString())
+                            .issuedTo(clusterService.getClusterName().value())
+                            .maxNodes(SELF_GENERATED_LICENSE_MAX_NODES)
+                            .issueDate(clock.millis())
+                            .type(License.LicenseType.BASIC)
+                            .expiryDate(BASIC_SELF_GENERATED_LICENSE_EXPIRATION_MILLIS);
+                        License selfGeneratedLicense = SelfGeneratedLicense.create(specBuilder, currentState.nodes());
+                        LicensesMetaData newLicensesMetadata = new LicensesMetaData(selfGeneratedLicense,
+                            currentLicensesMetadata == null ? null : currentLicensesMetadata.getMostRecentTrialVersion());
+                        mdBuilder.putCustom(LicensesMetaData.TYPE, newLicensesMetadata);
```

Comment:
Can we merge this code and the duplicate from `StartBasicClusterTask` into a single `generateNewBasicLicense()` method?